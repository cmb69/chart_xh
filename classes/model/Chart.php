<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Chart_XH.
 *
 * Chart_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Chart_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Chart_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Chart\Model;

use DOMDocument;
use DOMElement;
use DOMNode;
use Plib\Document2 as Document;
use Plib\DocumentStore2 as DocumentStore;

final class Chart implements Document
{
    public const TYPES = [
        "line",
        "bar",
        "horizontal-bar",
        "pie",
        "doughnut",
        "semi-pie",
        "semi-doughnut",
        "polar-area",
    ];

    private string $name;
    private string $caption;
    private string $type;
    private bool $transposed;
    private string $aspectRatio;
    /** @var list<string> */
    private array $labels = [];
    /** @var list<Dataset> */
    private array $datasets = [];

    public static function new(string $key): self
    {
        return new self(basename($key, ".xml"), "", "line", false, "3/2");
    }

    public static function fromString(string $contents, string $key): ?self
    {
        if ($contents === "") {
            return null;
        }
        $doc = new DOMDocument("1.0", "UTF-8");
        if (!$doc->loadXML($contents)) {
            return null;
        }
        if (!$doc->relaxNGValidate(__DIR__ . "/../../chart.rng")) {
            return null;
        }
        assert($doc->documentElement instanceof DOMElement);
        $chart = $doc->documentElement;
        $that = new self(
            basename($key, ".xml"),
            $chart->getAttribute("caption"),
            $chart->getAttribute("type"),
            (bool) $chart->getAttribute("transposed"),
            $chart->getAttribute("aspect-ratio")
        );
        foreach ($chart->childNodes as $childNode) {
            assert($childNode instanceof DOMNode);
            if ($childNode->nodeName === "label") {
                assert($childNode instanceof DOMElement);
                $that->labels[] = (string) $childNode->nodeValue;
            } elseif ($childNode->nodeName === "dataset") {
                assert($childNode instanceof DOMElement);
                $that->datasets[] = Dataset::fromXml($childNode);
            }
        }
        return $that;
    }

    public static function create(string $name, DocumentStore $store): self
    {
        $that = $store->create("$name.xml", self::class);
        assert($that instanceof self);
        return $that;
    }

    public static function read(string $name, DocumentStore $store): ?self
    {
        return $store->read("$name.xml", self::class);
    }

    public static function update(string $name, DocumentStore $store): ?self
    {
        return $store->update("$name.xml", self::class);
    }

    public function __construct(string $name, string $caption, string $type, bool $transposed, string $aspectRatio)
    {
        $this->name = $name;
        $this->caption = $caption;
        $this->type = $type;
        $this->transposed = $transposed;
        $this->aspectRatio = $aspectRatio;
    }

    public function name(): string
    {
        return $this->name;
    }

    public function caption(): string
    {
        return $this->caption;
    }

    public function type(): string
    {
        return $this->type;
    }

    public function transposed(): bool
    {
        return $this->transposed;
    }

    public function aspectRatio(): string
    {
        return $this->aspectRatio;
    }

    /** @return list<string> */
    public function labels(): array
    {
        return $this->labels;
    }

    /** @return list<Dataset> */
    public function datasets(): array
    {
        return $this->datasets;
    }

    public function setCaption(string $caption): void
    {
        $this->caption = $caption;
    }

    public function setType(string $type): void
    {
        $this->type = $type;
    }

    public function setTransposed(bool $transposed): void
    {
        $this->transposed = $transposed;
    }

    public function setAspectRatio(string $aspectRatio): void
    {
        $this->aspectRatio = $aspectRatio;
    }

    public function purgeLabels(): void
    {
        $this->labels = [];
    }

    public function addLabel(string $label): void
    {
        $this->labels[] = $label;
    }

    public function purgeDatasets(): void
    {
        $this->datasets = [];
    }

    public function addDataset(string $label, string $color): Dataset
    {
        return $this->datasets[] = new Dataset($label, $color);
    }

    public function toString(): string
    {
        $doc = new DOMDocument('1.0', 'UTF-8');
        $chart = $doc->createElement('chart');
        $chart->setAttribute("caption", $this->caption);
        $chart->setAttribute("type", $this->type);
        $chart->setAttribute("transposed", (string) $this->transposed);
        $chart->setAttribute("aspect-ratio", $this->aspectRatio);
        $doc->appendChild($chart);
        foreach ($this->labels as $label) {
            $elt = $doc->createElement("label");
            $elt->nodeValue = $label;
            $chart->appendChild($elt);
        }
        foreach ($this->datasets as $dataset) {
            $chart->appendChild($dataset->toXml($doc));
        }
        if (!$doc->relaxNGValidate(__DIR__ . "/../../chart.rng")) {
            return "";
        }
        $doc->formatOutput = true;
        return (string) $doc->saveXML();
    }
}
