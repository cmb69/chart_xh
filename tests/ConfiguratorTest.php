<?php

namespace Chart;

use ApprovalTests\Approvals;
use Chart\Model\Chart;
use PHPUnit\Framework\TestCase;
use Plib\DocumentStore2 as DocumentStore;

/** @medium */
class ConfiguratorTest extends TestCase
{
    private DocumentStore $store;

    public function setUp(): void
    {
        $this->store = new DocumentStore("./examples/");
    }

    private function sut(): Configurator
    {
        return new Configurator();
    }

    public function testConfiguresChart(): void
    {
        $chart = Chart::read("voting", $this->store);
        Approvals::verifyAsJson($this->sut()->configure($chart));
    }

    public function testConfiguresTransposedChart(): void
    {
        $chart = Chart::read("voting", $this->store);
        $chart->setTransposed(true);
        Approvals::verifyAsJson($this->sut()->configure($chart));
    }

    public function testConfiguresHorizontalBarChart(): void
    {
        $chart = new Chart("horz-bar", "Horizontal Bar Chart", "horizontal-bar", false, "3/2");
        $conf = $this->sut()->configure($chart);
        $this->assertSame("bar", $conf["type"]);
        $this->assertSame("y", $conf["options"]["indexAxis"]);
    }

    public function testConfiguresHalfPie(): void
    {
        $chart = new Chart("half-pie", "Semi-Pie Chart", "semi-pie", false, "3/2");
        $conf = $this->sut()->configure($chart);
        $this->assertSame("pie", $conf["type"]);
        $this->assertSame(-90, $conf["options"]["rotation"]);
        $this->assertSame(180, $conf["options"]["circumference"]);
    }

    public function testConfiguresHalfDoughnut(): void
    {
        $chart = new Chart("half-doughnut", "Semi-Doughnut Chart", "semi-doughnut", false, "3/2");
        $conf = $this->sut()->configure($chart);
        $this->assertSame("doughnut", $conf["type"]);
        $this->assertSame(-90, $conf["options"]["rotation"]);
        $this->assertSame(180, $conf["options"]["circumference"]);
    }

    public function testConfiguresPolarArea(): void
    {
        $chart = new Chart("polar-area", "Polar Area Chart", "polar-area", false, "3/2");
        $conf = $this->sut()->configure($chart);
        $this->assertSame("polarArea", $conf["type"]);
    }

    public function testNonRationalAspectRatioDefaultsToOne(): void
    {
        $chart = new Chart("polar-area", "Polar Area Chart", "polar-area", false, "17");
        $conf = $this->sut()->configure($chart);
        $this->assertSame(1.0, $conf["options"]["aspectRatio"]);
    }
}
