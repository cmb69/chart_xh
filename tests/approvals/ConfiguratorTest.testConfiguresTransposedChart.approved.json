{
    "data": {
        "datasets": [
            {
                "backgroundColor": [
                    "#ff0000",
                    "#000000",
                    "#00ff00",
                    "#ffff00",
                    "#0000ff",
                    "#ff00ff",
                    "#7f7f7f"
                ],
                "borderColor": [
                    "#ff0000",
                    "#000000",
                    "#00ff00",
                    "#ffff00",
                    "#0000ff",
                    "#ff00ff",
                    "#7f7f7f"
                ],
                "data": [
                    20.5,
                    33,
                    8.9,
                    10.7,
                    12.6,
                    9.2,
                    5
                ],
                "label": "2017"
            },
            {
                "backgroundColor": [
                    "#ff0000",
                    "#000000",
                    "#00ff00",
                    "#ffff00",
                    "#0000ff",
                    "#ff00ff",
                    "#7f7f7f"
                ],
                "borderColor": [
                    "#ff0000",
                    "#000000",
                    "#00ff00",
                    "#ffff00",
                    "#0000ff",
                    "#ff00ff",
                    "#7f7f7f"
                ],
                "data": [
                    25.7,
                    24.2,
                    14.7,
                    11.4,
                    10.4,
                    4.9,
                    8.7
                ],
                "label": "2021"
            },
            {
                "backgroundColor": [
                    "#ff0000",
                    "#000000",
                    "#00ff00",
                    "#ffff00",
                    "#0000ff",
                    "#ff00ff",
                    "#7f7f7f"
                ],
                "borderColor": [
                    "#ff0000",
                    "#000000",
                    "#00ff00",
                    "#ffff00",
                    "#0000ff",
                    "#ff00ff",
                    "#7f7f7f"
                ],
                "data": [
                    16.4,
                    28.6,
                    11.6,
                    4.3,
                    20.8,
                    8.8,
                    9.6
                ],
                "label": "2025"
            }
        ],
        "labels": [
            "SPD",
            "CDU\/CSU",
            "Gr\u00fcne",
            "FDP",
            "AfD",
            "Die Linke",
            "Sonstige"
        ]
    },
    "options": {
        "aspectRatio": 1.7777777777777777,
        "plugins": {
            "legend": {
                "labels": {
                    "boxHeight": 0,
                    "boxWidth": 0
                }
            }
        },
        "spanGaps": true
    },
    "type": "bar"
}